### DISCUSSION TITLE
pnpm `catalog:` protocol support (full support in Lerna-Lite v3.12)

### DISCUSSION THREAD

**Discussion by wassim-ben-amor** [ghiscoding/lerna-lite#901] [Ideas]
Hello,
[pnpm catalog](https://pnpm.io/catalogs) is officially released on `9.5.0`,
I wanted to know if there is a plan to support this new feature in `lerna-lite` 🙏 

I have checked the work done on https://github.com/lerna-lite/lerna-lite/pull/148, I think it should be similar.

Thank you,

**Comment by ghiscoding** [comment]
No I don't have any plans to support this and I don't think the goal would be to support every protocols that comes out. It makes sense to support the `workspace:` protocol but I really don't see why this new catalog should be supported which is a pnpm specific. On the other hand, `workspace:` was at least something that Yarn also supports and it is much more used and has been available for years (as opposed to this super new pnpm only catalog).

This new protocol also doesn't seem to be a 1-to-1 relationship like `workspace:` does, it's much more complex to support and I don't intend to support it at all. Like I said in previous paragraph, supporting every package managers little nuances is not the goal of this project

**Comment by JulianCataldo** [comment]
> I wanted to know if there is a plan to support this new feature in `lerna-lite` 🙏

I'm thinking of a quick fix relying on the [Life-cycle Scripts](https://github.com/lerna-lite/lerna-lite/tree/main/packages/version#lifecycle-scripts) during the `*version` commands.

In `preversion`, one could run a script to iterate over all workspaces package.json (e.g. reading lerna config),
replacing them with the relevant catalog versions.
Then undo all of this with `postversion`.

I didn't tested this. I'll try to figure out how to do that within Lerna's code preferably.

**Reply by wassim-ben-amor** [reply]
Thank you, 
using the `preversion` approach IMO will be too much work because you will need to detect the changed packages, especially that `lerna-lite changed` command will also fail as it also needs the dependency graph (so the same error will show up).

Following [pnpm catalogs docs](https://pnpm.io/catalogs), where we have defined the **default catalog** in `pnpm-workspace.yaml` file and probably added **named catalogs** in addition, this is the fix I have used maybe it can help others:
```diff
diff --git a/dist/package-graph/package-graph.js b/dist/package-graph/package-graph.js
index 0a3803b4d6ebf317346f975463f02fc69f783e1a..014314edaec3f72cc8622c791c6b44caffcdd290 100644
--- a/dist/package-graph/package-graph.js
+++ b/dist/package-graph/package-graph.js
@@ -1,3 +1,6 @@
+import fs from 'node:fs';
+import path from 'node:path';
+import { parse } from 'yaml';
 import npa from 'npm-package-arg';
 import { CyclicPackageGraphNode, PackageGraphNode, reportCycles } from './lib/index.js';
 import { ValidationError } from '../validation-error.js';
@@ -24,6 +27,7 @@ export class PackageGraph extends Map {
                 }
             }
         }
+        const catalogs = this.getWorkspacesCatalogs();
         this.forEach((currentNode, currentName) => {
             const graphDependencies = graphType === 'dependencies'
                 ? Object.assign({}, currentNode.pkg.optionalDependencies, currentNode.pkg.dependencies)
@@ -34,6 +38,19 @@ export class PackageGraph extends Map {
                 if (spec.startsWith('patch:')) {
                     spec = spec.replace(YARN_PATCH_PROTOCOL_REG_EXP, '$2');
                 }
+                const isCatalogSpec = /^catalog:/.test(spec);
+                if (isCatalogSpec) {
+                    spec = spec.replace(/^catalog:/, '');
+                    const catalogVersion = spec === '' || spec === 'default'
+                        ? catalogs.defaultCatalog[depName]
+                        : catalogs.namedCatalogs[spec]?.[depName];
+                
+                    if (catalogVersion) {
+                        spec = catalogVersion;
+                    } else {
+                        console.warn(`Warning: No version found in ${spec || 'default'} catalog for ${depName}`);
+                    }
+                }
                 const isWorkspaceSpec = /^workspace:/.test(spec);
                 let fullWorkspaceSpec;
                 if (isWorkspaceSpec) {
@@ -144,5 +161,16 @@ export class PackageGraph extends Map {
             node.localDependents.delete(candidateNode.name);
         });
     }
+    getWorkspacesCatalogs() {
+        const workspaceConfigPath = path.join(process.cwd(), 'pnpm-workspace.yaml');
+        const workspaceConfig = fs.existsSync(workspaceConfigPath)
+            ? parse(fs.readFileSync(workspaceConfigPath, 'utf8'))
+            : {};
+    
+        return {
+            defaultCatalog: workspaceConfig.catalog ?? {},
+            namedCatalogs: workspaceConfig.catalogs ?? {},
+        };
+    }
 }
 //# sourceMappingURL=package-graph.js.map
```

@ghiscoding, if you agree I can contribute with a PR to add the needed changes.

**Reply by ghiscoding** [reply]
@wassim-ben-amor  if this is all it takes to get `catalog:` protocol working in Lerna-Lite then yes for sure a PR contribution would be accepted, that is as long as it doesn't interfere with other protocols.

**Reply by ghiscoding** [reply]
@wassim-ben-amor is there still a possibility to get a contribution to add this new `catalog:` protocol support? A PR would certainly be welcome. Cheers

**Comment by ghiscoding** [comment]
The `catalog:` protocol is now available in Lerna-Lite [v3.11](https://github.com/lerna-lite/lerna-lite/releases/tag/v3.11.0) thanks to @wassim-ben-amor for providing the code and another thanks to @Mister-Hope for opening a PR for the feature 🚀. Also note that it comes with certain limitations as it was mentioned in the PR

> The `catalog:` protocol ([pnpm catalog](https://pnpm.io/catalogs)) can be recognized by Lerna-Lite. When publishing, they will be kept as is. If you need to bump the version of a package in a catalog, you will need to edit `pnpm-workspace.yaml` manually. So we suggest using workspace: protocol instead for workspace dependencies.

**Reply by ghiscoding** [reply]
Edit... so after giving a try myself, it looks like I misunderstood what the code above does. It seems that what this does it just to make `lerna version` happy by not crashing (great), however the code above doesn't fix `lerna publish` and users would instead have to use `pnpm publish` rather than `lerna publish`. 

It falls a bit short of what I was expecting and I would of course prefer to have a both lerna version & publish working (not just 1 command). That would however, require a bit more work, so we'll have to revisit this in the future since after trying the catalog, I think it does have some interesting concept and it would be nice to have it fully implemented (eventually)

**Comment by ghiscoding** [comment]
hey guys, I just published [v3.12.0](https://github.com/lerna-lite/lerna-lite/releases/tag/v3.12.0) which brings full [pnpm catalogs](https://pnpm.io/catalogs) support to both `lerna version` **and** `lerna publish` commands this time. I also replaced a few of my dependencies within the project to use `catalog:` protocol anywhere multiple packages use the same dependencies, and that was done purposely before publishing the new v3.12.0 release, so I can safely say that it works as intended for both commands :)
